### PR TITLE
EVP API ECDHE speed tool support

### PR DESCRIPTION
**Pull request by torben-hansen**
### Description of changes: 

Capture fips vs non-fips performance differences when using EVP API level to do ECDHE.

#### non-fips:
```
$ ./tool/bssl speed -filter "ECDH"                                                                                                        
Did 9250 ECDH P-224 operations in 1001966us (9231.9 ops/sec)
Did 17000 ECDH P-256 operations in 1007707us (16870.0 ops/sec)
Did 4213 ECDH P-384 operations in 1084221us (3885.7 ops/sec)
Did 3003 ECDH P-521 operations in 1098185us (2734.5 ops/sec)
Did 2519 ECDH secp256k1 operations in 1081613us (2328.9 ops/sec)
Did 9250 EVP ECDH P-224 operations in 1002814us (9224.0 ops/sec)
Did 17000 EVP ECDH P-256 operations in 1020140us (16664.4 ops/sec)
Did 3890 EVP ECDH P-384 operations in 1004690us (3871.8 ops/sec)
Did 2730 EVP ECDH P-521 operations in 1000664us (2728.2 ops/sec)
Did 2552 EVP ECDH secp256k1 operations in 1097794us (2324.7 ops/sec)
Did 30000 EVP ECDH X25519 operations in 1011202us (29667.7 ops/sec)
```

#### fips:
Notice the drop in the second part of the output.
```
$ ./tool/bssl speed -filter "ECDH"
Did 10175 ECDH P-224 operations in 1096756us (9277.4 ops/sec)
Did 17000 ECDH P-256 operations in 1006072us (16897.4 ops/sec)
Did 4125 ECDH P-384 operations in 1067952us (3862.5 ops/sec)
Did 2750 ECDH P-521 operations in 1001442us (2746.0 ops/sec)
Did 2618 ECDH secp256k1 operations in 1094624us (2391.7 ops/sec)
Did 3850 EVP ECDH P-224 operations in 1099314us (3502.2 ops/sec)
Did 6800 EVP ECDH P-256 operations in 1002298us (6784.4 ops/sec)
Did 1560 EVP ECDH P-384 operations in 1000189us (1559.7 ops/sec)
Did 1089 EVP ECDH P-521 operations in 1083282us (1005.3 ops/sec)
Did 1023 EVP ECDH secp256k1 operations in 1094584us (934.6 ops/sec)
Did 30000 EVP ECDH X25519 operations in 1012383us (29633.1 ops/sec)
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: tool/speed.cc [hunk 1]
```diff
@@ -1427,6 +1427,124 @@ static bool SpeedECDSA(const std::string &selected) {
...
+          EC_KEY_get0_public_key(EVP_PKEY_get0_EC_KEY(peer_key.get()))) ||
+        !EVP_PKEY_assign_EC_KEY(only_public_key_evp_pkey.get(), only_public_key_ec_key.release())) {
+      return false;
+    }
+    peer_key.reset(only_public_key_evp_pkey.release());
+  }
+
+  TimeResults results;
+  if (!TimeFunction(&results, [nid, &peer_key]() -> bool {
+    BM_NAMESPACE::UniquePtr<EVP_PKEY> my_key(evp_generate_key(nid));
```
Comment (nebeid) [inline tool/speed.cc:1504]: I suggest to add a comment: "Generate own key".

File: tool/speed.cc [hunk 2]
```diff
@@ -1427,6 +1427,124 @@ static bool SpeedECDSA(const std::string &selected) {
...
+
+    if (nid != NID_X25519) {
```
Comment (nebeid) [inline tool/speed.cc:1506]: - I suggest to add a comment: "Check peer key". 
- Should we check if it's a fips build and call `EC_KEY_check_fips()`?

Comment (nebeid) [reply tool/speed.cc:1506]: `EC_KEY_check_fips()` calls `EC_KEY_check_key()` and does a few other checks, which, hopefully shouldn't take more time. It is used for peer key and own key.

Comment (torben-hansen) [reply tool/speed.cc:1506]: SP 800-56Ar3 does say that... Need to fix...

File: tool/speed.cc [hunk 3]
```diff
@@ -1427,6 +1427,139 @@ static bool SpeedECDSA(const std::string &selected) {
...
+      // For the supported P NIST curves, the peer public key must be validated
+      // to ensure proper computation. FIPS has stronger requirements (cf.
+      // SP 800-56Ar3 5.6.2.2), so flip between the non-fips and fips version.
+      if (!ec_key_check_peer(EVP_PKEY_get0_EC_KEY(peer_key.get()))) {
```
Comment (nebeid) [inline tool/speed.cc:1522]: It turns out that checking the peer key is done from within `ECDH_compute_shared_secret()` on the peer key in all builds, and hence also from `EVP_PKEY_derive()`. This was introduced in [PR#358](https://github.com/aws/aws-lc/pull/358) for FIPS build only and then became part of the function in all builds in [PR#412](https://github.com/aws/aws-lc/pull/412), see `ecdh.c`.

File: tool/speed.cc [hunk 4]
```diff
@@ -1427,6 +1427,131 @@ static bool SpeedECDSA(const std::string &selected) {
...
+    BM_NAMESPACE::UniquePtr<EVP_PKEY> my_key(evp_generate_key(nid));
+
+#if defined(OPENSSL_BENCHMARK)
+    // For AWS-LC EVP_PKEY_derive() calls ECDH_compute_shared_secret() that
+    // performs the public key check.
+    if (nid != NID_X25519) {
+      // For the supported P NIST curves, the peer public key must be validated
+      // to ensure proper computation. FIPS has stronger requirements (cf.
+      // SP 800-56Ar3 5.6.2.2), so flip between the non-fips and fips version.
+      if (!EC_KEY_check_key(EVP_PKEY_get0_EC_KEY(peer_key.get()))) {
```
Comment (nebeid) [inline tool/speed.cc:1513]: Is the second part of the comment still applicable?
Also, I didn't check what OpenSSL does, you know that it doesn't check the peer key?

**Review by andrewhop — APPROVED** [review]

**Pull request merged**